pkg/proxy/runner/bounded_frequency_runner.go

```diff
@@ -0,0 +1,126 @@
+limitations under the License.
+*/
+
+package runner
+
+import (
+	"time"
+
+	"k8s.io/utils/clock"
+)
+
+type BoundedFrequencyRunner struct {
+	maxInterval time.Duration
+	minInterval time.Duration
+	fn          func()
+	queue       chan struct{}
+
+	clock            clock.Clock
+	maxIntervalTimer clock.Timer
+	minIntervalTimer clock.Timer
+}
+
+// NewBoundedFrequencyRunner creates and returns a new BoundedFrequencyRunner.
+// This runner manages the execution frequency of the provided function `fn`.
+//
+// The runner guarantees two properties:
+//  1. Minimum Interval (`minInterval`): At least `minInterval` must pass between
+//     the *completion* of one execution and the *start* of the next. Calls to
+//     `Run()` during this cooldown period are coalesced and deferred until the
+//     interval expires. This prevents burst executions.
+//  2. Maximum Interval (`maxInterval`): The function `fn` is guaranteed to run
+//     at least once per `maxInterval`, ensuring periodic execution even without
+//     explicit `Run()` calls (e.g., for refreshing state).
+//
+// Internally, the runner uses a buffered channel with a capacity of 1, acting
+// as a single-item queue. A dedicated goroutine (started by `Loop`) consumes
+// from this queue. This consumer loop triggers an execution of `fn` under two
+// conditions:
+//  1. An item is received from the queue (signaled by a `Run()` call).
+//  2. The `maxInterval` timer expires.
+//
+// After `fn` completes, the loop waits for the `minInterval` timer to expire
+// before it becomes eligible to execute `fn` again. This ensures the minimum
+// time between executions.
+//
+// Calls to `Run()` attempt to add an item to the queue. If the queue is already
+// full (meaning an execution is already pending), the `Run()` call is effectively
+// ignored (coalesced), as the existing item in the queue guarantees a future
+// execution.
+//
+// `maxInterval` must be greater than or equal to `minInterval`; otherwise,
+// this function will panic.
+func NewBoundedFrequencyRunner(fn func(), minInterval, maxInterval time.Duration) *BoundedFrequencyRunner {
+	if maxInterval < minInterval {
+		panic("maxInterval must be greater or equal than minInterval")
+	}
+
+	bfr := &BoundedFrequencyRunner{
+		maxInterval: maxInterval,
+		minInterval: minInterval,
+		queue:       make(chan struct{}, 1),
+		fn:          fn,
+		clock:       clock.RealClock{},
+	}
+	bfr.maxIntervalTimer = bfr.clock.NewTimer(bfr.maxInterval)
+	bfr.minIntervalTimer = bfr.clock.NewTimer(bfr.minInterval)
+	return bfr
+}
+
+func (bfr *BoundedFrequencyRunner) Loop(stop <-chan struct{}) {
+	defer close(bfr.queue)
+	defer bfr.maxIntervalTimer.Stop()
+	defer bfr.minIntervalTimer.Stop()
+
+	for {
+		select {
+		case <-stop:
+			return
+		case <-bfr.maxIntervalTimer.C(): // process work at least once every maxInterval
+		case <-bfr.queue: // process scheduled work
+		}
+
+		// execute
+		bfr.fn()
```

Comment:
Is it possible for two executions to occur if `Run()` is called simultaneously with `maxInterval`?

Reply:
golang `select` statement guarantees that only one wins, then the next execution will be wait for the `minSyncPeriod`